From c2f21687fe2c4e3a4651b25f9c73cfb58e90c8ce Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Thu, 3 Jul 2025 18:45:26 -0500
Subject: [PATCH 13/48] kernel32/tests: Test opening a directory with
 ReOpenFile().

(cherry picked from commit 6661ced770baf89c6192824977f874c47d62687c)

CW-Bug-Id: #26492
---
 dlls/kernel32/tests/file.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/dlls/kernel32/tests/file.c b/dlls/kernel32/tests/file.c
index bd862545726..8fdafa62f85 100644
--- a/dlls/kernel32/tests/file.c
+++ b/dlls/kernel32/tests/file.c
@@ -5088,6 +5088,22 @@ static void test_ReOpenFile(void)
     ret = DeleteFileA(filename);
     ok(ret, "failed to delete file, error %lu\n", GetLastError());
 
+    /* FILE_FLAG_BACKUP_SEMANTICS is allowed, but unlike with CreateFile(),
+     * it cannot be used to open a directory. */
+
+    ret = CreateDirectoryA(filename, NULL);
+    ok(ret == TRUE, "got error %lu\n", GetLastError());
+    file = CreateFileA(filename, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, 0);
+    ok(file != INVALID_HANDLE_VALUE, "got error %lu\n", GetLastError());
+    new = pReOpenFile(file, GENERIC_READ, FILE_SHARE_READ, FILE_FLAG_BACKUP_SEMANTICS);
+    todo_wine ok(new == INVALID_HANDLE_VALUE, "expected failure\n");
+    todo_wine ok(GetLastError() == ERROR_ACCESS_DENIED, "got error %lu\n", GetLastError());
+    CloseHandle(new);
+    CloseHandle(file);
+
+    ret = RemoveDirectoryA(filename);
+    ok(ret == TRUE, "got error %lu\n", GetLastError());
+
     file = CreateNamedPipeA("\\\\.\\pipe\\test_pipe", PIPE_ACCESS_DUPLEX, 0, 1, 1000, 1000, 1000, NULL);
     ok(file != INVALID_HANDLE_VALUE, "failed to create pipe, error %lu\n", GetLastError());
 
-- 
2.53.0

